lib/gc/Transforms/OneDNNGraphToLinalg.cpp

```diff
@@ -0,0 +1,280 @@
+#include "gc/Transforms/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Rewrite/FrozenRewritePatternSet.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir::onednn_graph;
+
+namespace mlir {
+namespace gc {
+#define GEN_PASS_DEF_CONVERTONEDNNGRAPHTOLINALG
+#include "gc/Transforms/Passes.h.inc"
+
+namespace {
+//===----------------------------------------------------------------------===//
+// Util funcs
+//===----------------------------------------------------------------------===//
+
+Value createBroadcastOperand(Location loc, PatternRewriter &rewriter,
+                             TensorType ty, Value op) {
+  auto opTy = dyn_cast<TensorType>(op.getType());
+  llvm::ArrayRef<int64_t> bcastShape = ty.getShape();
+  llvm::ArrayRef<int64_t> opShape = opTy.getShape();
+  int64_t diff = bcastShape.size() - opShape.size();
+
+  if (bcastShape.equals(opShape)) {
+    return op;
+  } else {
+    // get broadcast dimensions
+    llvm::SmallVector<int64_t> bcastDims;
+    for (int64_t i = 0; i < (int64_t)bcastShape.size(); i++) {
+      int64_t idxOp = i - diff;
+      if (idxOp < 0) {
+        bcastDims.push_back(i);
+      } else if (bcastShape[i] != opShape[idxOp]) {
+        bcastDims.push_back(i);
+      }
+    }
+    // create a new output tensor
+    Value initTensor =
+        rewriter.create<tensor::EmptyOp>(loc, bcastShape, ty.getElementType());
+    return rewriter
+        .create<linalg::BroadcastOp>(
+            /*location=*/loc,
+            /*inputs=*/op,
+            /*inits=*/initTensor,
+            /*dimensions=*/bcastDims)
+        .getResults()
+        .front();
+  }
+}
+
+typedef Value (*GetOperandFn)(Operation *, PatternRewriter &, TensorType);
+
+struct OriginalOperand {
+  template <unsigned I>
+  static Value getIdx(Operation *op, PatternRewriter &b, TensorType ty) {
+    if (I >= op->getNumOperands()) {
+      op->emitError("Index exceeds operand num.\n");
+      return nullptr;
+    }
+    return createBroadcastOperand(op->getLoc(), b, ty, op->getOperand(I));
+  }
+};
+
+struct ConstantOperand {
+  template <int64_t I>
+  static Value getConst(Operation *op, PatternRewriter &b, TensorType ty) {
+    const auto loc = op->getLoc();
+    if (llvm::isa<IntegerType>(ty.getElementType())) {
+      return b.create<arith::ConstantOp>( //
+          loc, DenseElementsAttr::get(ty, int64_t(I)));
+    } else if (llvm::isa<FloatType>(ty.getElementType())) {
+      return b.create<arith::ConstantOp>( //
+          loc, DenseElementsAttr::get(ty, float(I)));
+    } else {
+      op->emitError("Not a supported element type for constant.\n");
+      return nullptr;
+    }
+  }
+};
```

Comment:
These methods are stateless. What are they wrapped into a struct for? Is it just for logical grouping expecting more methods?

Reply:
Just for logical grouping, will add more, e.g. get constant operand from attrs like `onednn_graph.pow`